backend/api_v1/dns.go

```diff
@@ -78,252 +78,434 @@ type DNSResponse struct {
+		// Find the answer.
+		if len(msg.Answer) > 0 {
+			switch x := msg.Answer[0].(type) {
+			case *godns.NS:
+				return x.Ns, i, x.Hdr.Ttl, nil
+			case *godns.CNAME:
+				if recursionCount == 50 {
+					return "", 0, 0, fmt.Errorf("recursion limit reached on %s", hostname)
+				}
+				hostname = x.Target
+				recursionCount++
+				goto lookup
+			default:
+				return "", 0, 0, errors.New("invalid type for NS record")
 			}
-			return
 		}
+	}
+	if err != nil {
+		// Errored whilst trying to find NS record.
+		return "", 0, 0, err
+	}
 
-		// Get the type and hostname from the URL.
-		recordType := context.Param("recordType")
-		hostname := context.Param("hostname")
-		if !strings.HasSuffix(hostname, ".") {
-			hostname += "."
+	// Unable to find NS record.
+	log.Warn("unable to find NS record", zap.String("hostname", strings.Join(chunks, ".")))
+	return "", 0, 0, nil
+}
+
+func doDnsLookups(log *zap.Logger, dnsServer, recordType string, recursive bool, chunks []string) (map[string][]*DNSResponse, error) {
+	// Resolve the IP of the DNS server.
+	var addr string
+	setServerAddr := func() error {
+		rawAddr, err := net.ResolveIPAddr("ip", dnsServer)
+		if err != nil {
+			return err
 		}
+		addr = rawAddr.IP.String() + ":53"
+		return nil
+	}
+	if err := setServerAddr(); err != nil {
+		return nil, err
+	}
 
-		// Make the record type upper case.
-		recordType = strings.ToUpper(recordType)
-		originRecordType := recordType
+	// Keep going through chunks until we get a NS record.
+	initAddr := addr
+	oldDnsServer := dnsServer
+	host, i, ttl, err := findNameserverHostname(log, addr, chunks)
+	if err != nil {
+		return nil, err
+	}
+	if host == "" {
+		// Unable to find NS record.
+		log.Warn("unable to find NS record", zap.String("hostname", strings.Join(chunks, ".")))
+	} else {
+		// Turn it into the address.
+		dnsServer = host
+		if err = setServerAddr(); err != nil {
+			return nil, err
+		}
+	}
+
+	// If this was a NS lookup that is non-recursive, we have our result here.
+	if !recursive && recordType == "NS" {
+		if i == 0 {
+			// This means that the NS record was on the record specified.
+			return map[string][]*DNSResponse{
+				"NS": {
+					{
+						Type:      recordType,
+						TTL:       ttl,
+						Name:      strings.TrimRight(dnsServer, "."),
+						DNSServer: oldDnsServer,
+					},
+				},
+			}, nil
+		}
+
+		// No DNS records were found.
+		return map[string][]*DNSResponse{"NS": {}}, nil
+	}
 
-		// Defines the record types.
-		recordTypes := []string{recordType}
-		recordTypePacket, ok := godns.StringToType[recordType]
+	// Try to update the DNS server used here.
+	if err = setServerAddr(); err != nil {
+		return nil, err
+	}
+
+	// Define the record types.
+	recordTypes := []string{recordType}
+	if recordType == "ANY" {
+		// Not many DNS resolvers support this anymore, set it to literally all record types.
+		recordTypes = []string{"A", "AAAA", "CNAME", "MX", "NS", "PTR", "SOA", "TXT"}
```

Comment:
I just did some research into this, and I realised why I didn't include it. It is a bit of a special record type, only belonging to `_<service>._<protocol>`. For the most part, if you search it on a site, it is misleading since it will not be there.

Reply:
Yeah, that's a good point. Though I think we should still include it if the query matches something like `_.+\._.+\..+`.

Otherwise if you do a lookup for a specific DNS name (`_imaps._tcp.jimeh.me` for example) with type `ANY`, you'd not get any `SRV` records back.